nifi-testharness/src/main/java/org/apache/nifi/testharness/TestNiFiInstance.java

```diff
@@ -0,0 +1,620 @@
+            case START_FAILED:
+            case STOP_FAILED:
+            case STOPPED:
+
+                return Objects.requireNonNull(nifiVersion, "nifiVersion is null");
+
+            default:
+                throw new IllegalStateException(
+                        "NiFi version can only be retrieved after a successful installation, not in: "
+                                + currentState);
+        }
+    }
+
+    public void install() throws IOException {
+
+        currentState.checkCanTransition(State.INSTALLED);
+
+        File[] staleInstallations = placeholderNiFiHomeDir.listFiles((dir, name) -> name.startsWith("nifi-"));
+        if (staleInstallations != null) {
+            Arrays.stream(staleInstallations).forEach(TestNiFiInstance::deleteFileOrDirectoryRecursively);
+        }
+
+        Path tempDirectory = null;
+        try {
+            tempDirectory = Files.createTempDirectory("installable-flow");
+
+
+
+            LOGGER.info("Uncompressing NiFi archive {} to {} ...", nifiBinaryZip, placeholderNiFiHomeDir);
+
+            Zip.unzipFile(nifiBinaryZip, placeholderNiFiHomeDir, new Zip.StatusListenerAdapter() {
+                @Override
+                public void onUncompressDone(ZipEntry ze) {
+                    LOGGER.debug("Uncompressed {}", ze.getName());
+                }
+            });
+
+            LOGGER.info("Uncompressing DONE");
+
+            File actualNiFiHomeDir = getActualNiFiHomeDir(placeholderNiFiHomeDir);
+
+            nifiVersion = getNiFiVersion(actualNiFiHomeDir);
+
+            currentState = State.INSTALLED;
+
+            File installableFlowFile = createInstallableFlowFile(tempDirectory);
+
+            validateNiFiVersionAgainstFlowVersion(nifiVersion, installableFlowFile);
+
+            FileUtils.createSymlinks(placeholderNiFiHomeDir, actualNiFiHomeDir);
+
+            installFlowFile(installableFlowFile);
+        } catch (Exception e) {
+
+            currentState = State.INSTALLATION_FAILED;
+
+            throw new RuntimeException("Installation failed: " + e.getMessage(), e);
+
+        } finally {
+            if (tempDirectory != null) {
+                FileUtils.deleteDirectoryRecursive(tempDirectory);
+            }
+        }
+
+        currentState = State.FLOW_INSTALLED;
+    }
+
+    private File createInstallableFlowFile(Path tempDirectory) throws IOException {
+
+        File flowXmlFile = new File(tempDirectory.toFile(), "flow.xml");
+
+        if (editCallback == null) {
+            Files.copy(flowXml.toPath(), flowXmlFile.toPath());
+        } else {
+            if (editCallback instanceof TestNiFiInstanceAware) {
+                ((TestNiFiInstanceAware)editCallback).setTestNiFiInstance(this);
+            }
+
+            XmlUtils.editXml(flowXml, flowXmlFile, editCallback);
+        }
+
+        return flowXmlFile;
+    }
+
+    private void installFlowFile(File fileToIncludeInGz) throws IOException {
+        Zip.gzipFile(fileToIncludeInGz, flowXmlGz);
```

Comment:
@markap14 @bbende Can one of you take a look here at how he's building up the test flow.xml.gz? I have a feeling he's making a bad assumption because the synchronizer code is blowing up in my log file in a place where it never does outside of this use case.

Reply:
NM, documentation problem! (Or lack thereof)

Reply:
@bbende @markap14 In case you didn't get an update, no need here.